### ISSUE TITLE
Solcjs doesn't resolve paths

### ISSUE THREAD

**Issue by PascalPrecht** [ethereum/solc-js#377]
There seems to be another difference in how `solcjs` behaves compared to `solc`.

With a Smart Contract like this (let's call it `SimpleStorage.sol`):

```
pragma solidity ^0.5.0;

import "/Users/pascalprecht/projects/vibranium-dapp/contracts/test.sol";
import "/Users/pascalprecht/projects/vibranium-dapp/contracts/foo.sol";

contract SimpleTestContract {
  uint public storedData;

  constructor(uint initialValue) public {
    storedData = initialValue;
  }

  function set(uint x) public {
    storedData = x;
  }

  function get() public view returns (uint retVal) {
    return storedData;
  }
}
```
Notice the absolute import paths ^,  I can successfully compile this code using `solc` with a command like:

```
solc --abi --bin --overwrite -o /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/artifacts \
   /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/SimpleStorage.sol \
   /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/foo.sol
```
Notice how the applied paths point to the same files as the import paths in `SimpleStorage.sol` even though, the paths are not exactly the same. `solc` is smart enough to resolve those paths properly.

With `solcjs` this is different. Running this command:

```
solcjs --abi --bin -o /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/artifacts \
  /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/SimpleStorage.sol \
  /Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/foo.sol
```
I'll get **ParseErrors** with the following message:

```
/Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/SimpleStorage.sol:3:1: ParserError: Source "/Users/pascalprecht/projects/vibranium-dapp/contracts/test.sol" not found: File import callback not supported
import "/Users/pascalprecht/projects/vibranium-dapp/contracts/test.sol";
^----------------------------------------------------------------------^

/Users/pascalprecht/projects/vibranium/cli/../../vibranium-dapp/contracts/SimpleStorage.sol:4:1: ParserError: Source "/Users/pascalprecht/projects/vibranium-dapp/contracts/foo.sol" not found: File import callback not supported
import "/Users/pascalprecht/projects/vibranium-dapp/contracts/foo.sol";
^---------------------------------------------------------------------^
```

It turns out that this is actually **not** a ParseError. The paths are fine but `solcjs` will always report paths as a parse error if it "can't find" them. The reason it can't find them, is because the paths applied to the `solcjs` command differ from the import paths (even though they point at the same files).

So to make this work, the paths need be always exact when they are absolute:

```
solcjs --abi --bin -o /Users/pascalprecht/projects/vibranium-dapp/artifacts \
  /Users/pascalprecht/projects/vibranium-dapp/contracts/SimpleStorage.sol \
  /Users/pascalprecht/projects/vibranium-dapp/contracts/foo.sol
```

^ This compiles completely fine.

Given that `solc` seems to be smart enough that `/some/relative/../path` and `/some/path` are the same, I'd expect `solcjs` to understand this as well.

**In summary**:
- `solcjs` should probably not report something as a **ParseError** when it's in fact a path resolution error
- If it does report an error, it should exit the process with an error code (right now its 0 even with the "parse errors")
- `solcjs` should most likely understand and resolve different paths that point to the same file

I hope this makes sense!

**Comment by chriseth** [comment]
Resolving imports is part of the parsing stage (as opposed to the analysis stage). That's why it is a parser error. We could introduce a new error type if you like.

solcjs is not meant to be a full replacement for solc (the binary release of Solidity) and importing non-specified files (including relative path resolution) is one feature supported by solc but not by solcjs.

Would sanitizing the input paths already help you?

**Comment by PascalPrecht** [comment]
Hi @chriseth thanks for getting back to me on this one.

> Resolving imports is part of the parsing stage (as opposed to the analysis stage). That's why it is a parser error. We could introduce a new error type if you like.

Fair enough. It did throw me off tho because a parse error, to me, would be something like wrong syntax etc. (something that couldn't be parsed). I would argue that the imports above are perfectly fine from a parsing perspective, but I'm also no expert in this so maybe this is indeed expected behaviour.

Otherwise, a new error type would be helpful.

> solcjs is not meant to be a full replacement for solc (the binary release of Solidity) and importing non-specified files (including relative path resolution) is one feature supported by solc but not by solcjs.

I figured. If you say this won't be supported in the future, I'm happy to close this issue. Maybe this should be added somewhere in the docs/readme in a "differences" section or something.

> Would sanitizing the input paths already help you?

That's what I'm doing now yes. I have to canonicalize them to get rid off `../` etc, so the import paths used within the source file match exactly what's given as input to make this work.

**Comment by PascalPrecht** [comment]
Oh before I forget, can we make `solcjs` at least exit with an error code? Cause right now it's exiting with 0 making it a little inconvenient when composing with other programs.
